### PR TITLE
Add firestore 2nd gen triggers

### PR DESCRIPTION
**Pull request by colerogers**
Add support for 2nd gen firestore triggers!
```
import {
  onDocumentWritten,
  onDocumentCreated,
  onDocumentUpdated,
  onDocumentDeleted,
  Change
  FirestoreEvent
} from "firebase-functions/v2/firestore";

export const a = onDocumentWritten("foo/{bar}", (event: FirestoreEvent<Change<DocumentSnapshot>, { bar: string }>) => {});

export const b = onDocumentCreated("foo/{bar}", (event: FirestoreEvent<QueryDocumentSnapshot, { bar: string }>) => {});

export const c = onDocumentUpdated("foo/{bar}", (event: FirestoreEvent<Change<QueryDocumentSnapshot>, { bar: string }>) => {});

export const d = onDocumentDeleted("foo/{bar}", (event: FirestoreEvent<QueryDocumentSnapshot, { bar: string }>) => {});
```

### CODE REVIEW

File: src/v2/providers/firestore.ts [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+  document: string;
+  /**
+   * An object containing the values of the path patterns.
+   * Only named capture groups will be populated - {key}, {key=*}, {key=**}
+   */
+  params: Params;
+}
+
+/** DocumentOptions extend EventHandlerOptions with provided document and optional database and namespace.  */
+export interface DocumentOptions<Document extends string = string> extends EventHandlerOptions {
```
Comment (taeold) [inline src/v2/providers/firestore.ts:104]: TS Q: I'm not sure why we need `Document extends string` here. Isn't Document always going to be string anyway?

I see similar pattern in our v1 code (`Path extends string`) and I'm not sure why that's there too 🤔

**Review by taeold — APPROVED** [review]

File: package.json
[the remark this answers is not in the document]
Comment (colerogers) [reply package.json:1]: Added!

File: src/common/providers/firestore.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }
+  const proto = {
+    fields: value?.fields || {},
+    createTime: dateToTimestampProto(value?.createTime),
+    updateTime: dateToTimestampProto(value?.updateTime),
+    name: value?.name || resource,
+  };
+  return proto;
+}
+
```
[the remark this answers is not in the document]
Comment (colerogers) [reply src/common/providers/firestore.ts:54]: Yep, fixed!

File: src/v2/providers/firestore.ts [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+  } else if (event.datacontenttype.includes("application/json")) {
+    return createSnapshotFromJson(
+      event.data,
+      event.source,
+      (event.data as RawFirestoreData).value?.createTime,
+      (event.data as RawFirestoreData).value?.updateTime
+    );
+  } else {
+    logger.error("Cannot determine payload type, failing out.");
+    throw Error("Error: Cannot parse event payload.");
```
[the remark this answers is not in the document]
Comment (colerogers) [reply src/v2/providers/firestore.ts:311]: fixed!

File: src/v2/providers/firestore.ts (hunk 1 above)
Comment (colerogers) [reply src/v2/providers/firestore.ts:104]: It's used so we can enforce strong typing in our code - https://stackoverflow.com/questions/57342018/extending-a-string-in-typescript

I think Thomas was the one that actually put the Path stuff in during a refactor.

**Comment by patrikjuvonen** [comment]
This fixed issue #1213

**Pull request merged**